File: MainSequential.java
public class MainSequential {
    public static void main(String[] args) {
        int rows = 100;  // Default grid size
        int cols = 100;
        int steps = 100;
        long seed = 12345;
        String input_filename = "";
        String output_file = "weather_simulation.csv";


        if (args.length > 0) {
            rows = Integer.parseInt(args[0]);
        }
        if (args.length > 1) {
            cols = Integer.parseInt(args[1]);
        }
        if (args.length > 2) {
            steps = Integer.parseInt(args[2]);
        }
        if (args.length > 3) {
            seed = Long.parseLong(args[3]);
        }
        if (args.length > 4) {
            input_filename = args[4];
        }

        Grid grid = new Grid(rows, cols, seed);

        if (input_filename.isEmpty()) {
            grid.initializeCells();
        } else {
            grid.loadGridFromCSV(input_filename);
        }


        for (int i = 0; i < steps; i++) {
            System.out.println("Step " + (i + 1));
            grid.update();  // Serial update
            // grid.displayGrid(); // Optional depending on if we want to do this or not.
            grid.saveGridToCSV(output_file, i + 1);
        }
    }
}

File: MainSequential.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include "Grid.h"

int main(int argc, char *argv[]) {
    // Default values matching Java implementation
    int rows = 100;
    int cols = 100;
    int steps = 100;

    // Parse command line arguments (matching Java's argument parsing)
    if (argc > 1) {
        rows = atoi(argv[1]);
    }
    if (argc > 2) {
        cols = atoi(argv[2]);
    }
    if (argc > 3) {
        steps = atoi(argv[3]);
    }

    // Initialize random number generator
    srand(time(NULL));

    // Add timing variables
    clock_t start_time = clock();
    time_t start_wall = time(NULL);

    // Create and initialize grid
    Grid* grid = Grid_new(rows, cols);
    if (grid == NULL) {
        fprintf(stderr, "Failed to create grid\n");
        return 1;
    }
    Grid_initialize_cells(grid);

    const char* filename = "weather_simulation.csv";
    FILE* file = fopen(filename, "w");
    fclose(file);

    // Main simulation loop
    for (int i = 0; i < steps; i++) {
        printf("Step %d\n", i + 1);
        Grid_update(grid);  // Serial update
        // Grid_display_grid(grid);  // Optional, commented out as in Java
        Grid_save_grid_to_csv(grid, filename, i + 1);
    }

    // Calculate and print timing results
    clock_t end_time = clock();
    time_t end_wall = time(NULL);
    
    double cpu_time_used = ((double) (end_time - start_time)) / CLOCKS_PER_SEC;
    double wall_time_used = difftime(end_wall, start_wall);
    
    printf("CPU Time used: %.2f seconds\n", cpu_time_used);
    printf("Wall clock time: %.2f seconds\n", wall_time_used);

    // Clean up
    Grid_destroy(grid);

    return 0;
} 